File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/G._Банкомат/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(M)
    """
    m: int = int(sys.stdin.readline().rstrip())
    n: int = int(sys.stdin.readline().rstrip())
    coins: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    dp: list[int] = [0] * (m + 1)
    dp[0] = 1
    for i in range(n):
        for exchange in range(coins[i], m + 1):
            dp[exchange] += dp[exchange - coins[i]]
    sys.stdout.write(f'{dp[m]}\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/G._Банкомат/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(M)
    */
    int m;
    std::scanf("%d", &m);
    int n;
    std::scanf("%d", &n);
    std::vector<int> coins (n);
    for (int i = 0; i < n; ++i) std::cin >> coins[i];
    std::vector<int> dp (m + 1, 0);
    dp[0] = 1;
    for (int i = 0; i < n; ++i) {
        for (int exchange = coins[i]; exchange <= m; ++exchange) {
            dp[exchange] += dp[exchange - coins[i]];
        }
    }
    std::cout << dp[m] << "\n";
}


int main() {
    solution();
}